pkg/KubeArmorOperator/internal/controller/resources.go

```diff
@@ -130,12 +135,19 @@ func genRuntimeVolumes(runtime, runtimeSocket, runtimeStorage string) (vol []cor
 			})
 
 			storageLocation = common.RuntimeStorageLocation[runtime]
-			volMnt = append(volMnt, corev1.VolumeMount{
+			runtimeStorageVolMnt := corev1.VolumeMount{
 				Name:             runtime + "-storage",
 				MountPath:        storageLocation,
 				MountPropagation: &common.HostToContainerMountPropagation,
 				ReadOnly:         true,
-			})
+			}
+
+			// For some reason(s) readonly containerd runtime storage volume mount failed to
+			// create shim task on k0s.
```

Comment:
Can we understand the reasoning? I don't like that we have a special handling for k0s

Reply:
k0s containerd runtime state directory structure:
```shell
root@k0s:~# ls -l /run/k0s
total 40
drwx--x--x 5 root           root  100 Oct 29 09:55 containerd
-rw-r--r-- 1 root           root 4298 Oct 29 09:15 containerd-cri.toml
-rw-r--r-- 1 root           root    5 Oct 29 09:15 containerd.pid
srw-rw---- 1 root           root    0 Oct 29 09:15 containerd.sock
srw-rw---- 1 root           root    0 Oct 29 09:15 containerd.sock.ttrpc
-rw------- 1 root           root 4389 Oct 29 09:15 k0s.yaml
drwxr-xr-x 2 kube-apiserver root   60 Oct 29 09:15 kine
-rw-r--r-- 1 root           root    5 Oct 29 09:15 kine.pid
-rw-r--r-- 1 root           root    5 Oct 29 09:15 kube-apiserver.pid
-rw-r--r-- 1 root           root    5 Oct 29 09:15 kube-controller-manager.pid
-rw-r--r-- 1 root           root    5 Oct 29 09:15 kubelet.pid
-rw-r--r-- 1 root           root    5 Oct 29 09:15 kube-scheduler.pid
srwxr-xr-x 1 root           root    0 Oct 29 09:15 status.sock
root@k0s:~#

root@k0s:~# ls -l /run/k0s/containerd
total 0
drwxr-xr-x 4 root root 80 Oct 29 09:16 io.containerd.grpc.v1.cri
drwx--x--x 2 root root 40 Oct 29 09:15 io.containerd.runtime.v1.linux
drwx--x--x 3 root root 60 Oct 29 09:15 io.containerd.runtime.v2.task
root@k0s:~#
```

k3s containerd runtime state directory structure:
```shell
root@k3s:~# ls -l /run/k3s
total 0
drwx--x--x 5 root root 140 Oct 29 09:48 containerd

root@k3s:~# ls -l /run/k3s/containerd
total 0
srw-rw---- 1 root root  0 Oct 29 09:25 containerd.sock
srw-rw---- 1 root root  0 Oct 29 09:25 containerd.sock.ttrpc
drwxr-xr-x 4 root root 80 Oct 29 09:26 io.containerd.grpc.v1.cri
drwx--x--x 2 root root 40 Oct 29 09:25 io.containerd.runtime.v1.linux
drwx--x--x 3 root root 60 Oct 29 09:26 io.containerd.runtime.v2.task
root@k3s:~#
```

**The points to note here is:**
- By default k3s containerd stores all its state at `/run/k3s/containerd` path including the socket.
- By [default](https://docs.k0sproject.io/v1.28.2+k0s.0/runtime/#user-managed-containerd-configuration) k0s containerd stores its socket at `/run/k0s` path and other data at `/run/k0s/containerd` path.

**The reason that readonly containerd state volume-mount is working with k3s and not with k0s is:**

- We mount the k3s containerd state directory `run/k3s/containerd` at `/run/containerd` mount-path which also contains the socket so mounting the containerd socket at `/var/run/containerd/containerd.sock` mount-path has no effect.

- k0s stores its socket and other data at different paths. Therefore, when we mount the `/run/k0s/containerd` directory at `/run/containerd` mount-path as readonly and try to mount the socket at `/var/run/containerd/containerd.sock` mount-path, it will fail for obvious reason.

**Possible solutions I can think of:**

- Mount runtime storage volume without readonly flag for k0s as proposed in this PR.
- Mount containerd socket at `/run/containerd.sock` or at some other mount-path irrespective of k3s, k0s, etc. and no need to make any changes in the mount-path for runtime storage volume. This requires changes to the containerd socket map [here](https://github.com/kubearmor/KubeArmor/blob/5d1b3159a64ec1ac61a539b82d2de3534f74ba45/KubeArmor/common/common.go#L384) and [here](https://github.com/kubearmor/KubeArmor/blob/5d1b3159a64ec1ac61a539b82d2de3534f74ba45/pkg/KubeArmorOperator/common/defaults.go#L186).